# Property's Sale Price Prediction Using Linear Regression
Dataset: https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data?select=train.csv

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn import datasets, linear_model, metrics

In [4]:
df = pd.read_csv('C:/Users/rosyi/Videos/Machine Learning/Day 13. Supervised Learning - Linear Regression/train.csv')
df

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

#### Cleansing Data

In [5]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

Terdapat beberapa column yang memiliki nilai Nan, hal yang perlu dilakukan adalah dengan menghilangkan column dengan jumlah Nan lebih dari 50% dan menggunakan nilai mean/nedian untuk column yang memiliki jumlah Nan yang kurang dari 50%

In [7]:
# Menghilangkan column yang memiliki Nan lebih dari 50%
df.drop(['Alley', 'FireplaceQu', 'PoolQC', 'Fence', 'MiscFeature'], axis = 'columns', inplace=True)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 76 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   LotShape       1460 non-null   object 
 7   LandContour    1460 non-null   object 
 8   Utilities      1460 non-null   object 
 9   LotConfig      1460 non-null   object 
 10  LandSlope      1460 non-null   object 
 11  Neighborhood   1460 non-null   object 
 12  Condition1     1460 non-null   object 
 13  Condition2     1460 non-null   object 
 14  BldgType       1460 non-null   object 
 15  HouseStyle     1460 non-null   object 
 16  OverallQual    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  YearBuil

In [9]:
# Mengisi column numerikal yang memiliki Nan dengan median
df[['LotFrontage','MasVnrArea', 'GarageYrBlt']] = df[['LotFrontage','MasVnrArea', 'GarageYrBlt']].fillna(df[['LotFrontage','MasVnrArea', 'GarageYrBlt']].median()) 

# Mengisi column categorical dengan modus.
df[['MasVnrType','BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']] = df[['MasVnrType','BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']].apply(lambda x: x.fillna(x.value_counts().index[0]))

In [10]:
from pandas import isnull

np.where(df == isnull)

# Data sudah tidak memiliki nilai NaN yang berarti sudah clean dan siap untuk digunakan ke tahap selanjutnya.

(array([], dtype=int64), array([], dtype=int64))

#### One-Hot Encoding

In [11]:
OHD = pd.get_dummies(df, columns = ['MSZoning', 'Street', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'SaleType','SaleCondition'])
OHD

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1455  1456          60         62.0     7917            6            5   
1456  1457          20         85.0    13175            6            6   
1457  1458          70         66.0     9042            7            9   
1458  1459          20         68.0     9717            5            6   
1459  1460          20         75.0     9937            5            6   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLw  \
0          2003          2003       196.0         706  ...               0   
1          1976          1976         0.0         978  ...               0   
2          2001          2002       162.0         486  ...               0   
3          1915          1970         0.0         216  ...               0   
4          2000          2000       350.0         655  ...               0   
...         ...           ...         ...         ...  ...             ...   
1455       1999          2000         0.0           0  ...               0   
1456       1978          1988       119.0         790  ...               0   
1457       1941          2006         0.0         275  ...               0   
1458       1950          1996         0.0          49  ...               0   
1459       1965          1965         0.0         830  ...               0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
0                0             0            1                      0   
1                0             0            1                      0   
2                0             0            1                      0   
3                0             0            1                      1   
4                0             0            1                      0   
...            ...           ...          ...                    ...   
1455             0             0            1                      0   
1456             0             0            1                      0   
1457             0             0            1                      0   
1458             0             0            1                      0   
1459             0             0            1                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                         0                     0                     0   
1                         0                     0                     0   
2                         0                     0                     0   
3                         0                     0                     0   
4                         0                     0                     0   
...                     ...                   ...                   ...   
1455                      0                     0                     0   
1456                      0                     0                     0   
1457                      0                     0                     0   
1458                      0                     0                     0   
1459                      0                     0                     0   

      SaleCondition_Normal  SaleCondition_Partial  
0                        1                      0  
1                        1                      0  
2                        1                      0  
3                        0                      0  
4                        1                      0  
...                    ...                    ...  
1455                     1                      0  


#### Create X and Y variable for training and testing

In [12]:
X = OHD.loc[:, OHD.columns != 'SalePrice']
y = OHD['SalePrice']

#### Splitting X and y into training and testing sets

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4,
                                                    random_state=1)

#### Train Linear Regression with X_train and y_train

In [14]:
from sklearn.linear_model import LinearRegression

# create linear regression object
reg = linear_model.LinearRegression()
 
# train the model using the training sets
reg.fit(X_train, y_train)
 
# regression coefficients
print('Coefficients: ', reg.coef_)

# regression intercept
print('Intercept: ', reg.intercept_)

Coefficients:  [ 2.80083267e+00 -1.57396024e+02  5.13663610e+01  1.08310570e+00
  4.88790458e+03  6.07457676e+03  2.39259088e+02  3.33935675e+01
  1.48043563e+01  1.91606866e+01  1.02564681e+00 -3.01253820e+00
  1.71738282e+01  1.74963307e+01  2.64050267e+01 -6.25099895e+00
  3.76503488e+01 -7.29837917e+02  1.30130930e+03  3.73953137e+03
  1.87301635e+03 -1.73578156e+03 -1.17061033e+04 -9.18486652e+01
  2.56138195e+03  2.32370643e+01  4.43011323e+03  1.29127168e+01
  6.37790424e+00  2.57693421e+01 -3.56895758e+00  1.94948445e+01
  1.66634377e+01  4.03498298e+01  6.58252162e-01 -2.96096559e+02
 -7.15920701e+02 -3.68863480e+04  1.76955975e+04  1.74019435e+04
  8.40114440e+02  9.48692566e+02  1.47002986e+03 -1.47002986e+03
  2.59411114e+03  1.28083660e+03 -7.32873844e+03  3.45379070e+03
  1.84008245e+02  1.13466687e+04 -1.64317525e+04  4.90107554e+03
  2.45846356e+04 -2.45846356e+04  2.06511350e+03  7.06481354e+03
 -5.60361045e+02 -7.59034460e+03 -9.79221396e+02  2.53135656e+04
  2.664780

#### Evaluate with 3 metrics

In [15]:
y_pred = reg.predict(X_test)

# r^2 score
from sklearn.metrics import r2_score
print('r^2 Score: ', r2_score(y_test,y_pred))

# MAE
from sklearn.metrics import mean_absolute_error
print('MAE : ',mean_absolute_error(y_test, y_pred))

# RMSE
from sklearn.metrics import mean_squared_error
print('RMSE : ',mean_squared_error(y_test, y_pred, squared=False))

r^2 Score:  0.7133951119985935
MAE :  20280.566852174532
RMSE :  43523.98323232034


#### Conclusion

- nilai r^2 dari model terhadap data adalah sebesar 0.71 atau 71% yang menandakan bahwa pemodelan yang dibuat mirip sebesar 71% terhadap data yang ada dan dapat dikatakan baik. 
- Nilai MAE menyatakan besaran error rata-rata pada model prediksi, dimana MAE yang didapatkan adalah sebesar 20280.566852174532 yang berarti kemungkinan rata-rata error pada mooel dari data aslinya adalah +20280.566852174532 dan -20280.566852174532
- RMSE adalah skor kuadrat yang mengkur besar error rata-rata, dalam kasus ini didapatkan RMSE sebesar +43523.98323232034 dan -43523.98323232034 dari data aslinya.